virtcontainers/netmon/netmon.c

```diff
@@ -0,0 +1,1301 @@
+		free(iface_list[nif.idx].hw_addr);
+		iface_list[nif.idx].hw_addr = strdup(nif.hw_addr);
+	}
+
+	/* MTU */
+	if (nif.mtu) {
+		free(iface_list[nif.idx].mtu);
+		iface_list[nif.idx].mtu = strdup(nif.mtu);
+	}
+}
+
+/*
+ * Update interface subroutine that can be used both from add_interface()
+ * and update_interface().
+ */
+int _update_interface(const struct nlmsghdr *nh, struct iface nif,
+		const char *sandbox_id, const char* runtime_path) {
+	int ret;
+
+	/* 
+	 * Call into the runtime with the proper command:
+	 * # kata-runtime upd-net-if --name ...
+	 */
+	printf("# %s upd-net-if --name %s --hw-addr %s --mtu %s\n",
+		runtime_path, iface_list[nif.idx].name,
+		iface_list[nif.idx].hw_addr, iface_list[nif.idx].mtu);
+
+	/* If the exec terminated properly, let's update iface_list */
+	update_iface_list(nif);
+
+	if (nh->nlmsg_type == RTM_NEWADDR) {
+		ret = insert_ip_addr(nif.idx,
+				     (const char*)nif.ip_addrs->addr,
+				     nif.ip_addrs->family);
+	} else if (nh->nlmsg_type == RTM_DELADDR) {
+		ret = delete_ip_addr(nif.idx,
+				     (const char*)nif.ip_addrs->addr);
+	}
+
+	free_iface(&nif);
+
+	return ret;
+}
+
+/*
+ * Add interface.
+ */
+int add_interface(const struct nlmsghdr *nh, const char *sandbox_id,
+		const char* runtime_path)
+{
+	char *params[3];
+	struct iface nif = parse_ifinfomsg(nh);
+
+	if (iface_list[nif.idx].idx != INIT_IF_IDX) {
+		if (!iface_changed(nif, iface_list[nif.idx])) {
+			printf("Interface %s didn't change\n", nif.name);
+			return 0;
+		}
+
+		return _update_interface(nh, nif, sandbox_id, runtime_path);
+	}
+
+	/* 
+	 * Call into the runtime with the proper command:
+	 * # kata-runtime add-net-if --name ...
+	 */
+	printf("# %s add-net-if --name %s --hw-addr %s --mtu %s\n",
+			runtime_path, nif.name, nif.hw_addr, nif.mtu);
+
+	params[0] = (char *) runtime_path;
+	asprintf(&(params[1]), "add-net-if --name %s --hw-addr %s --mtu %s",
+			nif.name, nif.hw_addr, nif.mtu);
+	params[2] = NULL;
+
+	//fork_runtime_call(params);
+
+	/* If the exec terminated properly, let's add it to iface_list */
+	add_iface_to_list(nif);
+
+	free(params[1]);
+	free_iface(&nif);
+
+	return 0;
+}
+
+/*
+ * Update interface.
+ */
+int update_interface(const struct nlmsghdr *nh, const char *sandbox_id,
+		const char* runtime_path)
+{
+	struct iface nif = parse_ifaddrmsg(nh);
+
+	return _update_interface(nh, nif, sandbox_id, runtime_path);
+}
+
+/*
+ * Delete interface.
+ */
+int delete_interface(const struct nlmsghdr *nh, const char *sandbox_id,
+		const char* runtime_path)
+{
+	struct iface nif = parse_ifinfomsg(nh);
+
+	/* 
+	 * Call into the runtime with the proper command:
+	 * # kata-runtime del-net-if --name ...
+	 */
+	printf("# %s del-net-if --name %s\n", runtime_path, nif.name);
+
+	/* If the exec terminated properly, let's delete it from iface_list */
+	delete_iface_from_list(nif.idx);
+
+	free_iface(&nif);
+
+	return 0;
+}
+
+/*
+ * Add route.
+ */
+int add_route(const struct nlmsghdr *nh, const char *sandbox_id,
```

Comment:
For this kind of change, I would insist that either we make change to network config via the sandbox network hotplug API, or we put this binary in the cli directory. I do not think we should make virtcontainers call the kata-runtime binary directly.

Reply:
See my comment [here](https://github.com/kata-containers/runtime/pull/534#discussion_r206663151), let me know what you think about it !